private/pkg/manifest/module.go

```diff
@@ -245,3 +229,71 @@ func NewMemoryBlobFromReaderWithDigester(content io.Reader, digester Digester) (
 		content: contentInMemory.Bytes(),
 	}, nil
 }
+
+// BlobEqual returns true if blob a is the same as blob b. The digest is
+// checked for equality and the content bytes compared.
+//
+// An error is returned if an unexpected I/O error occurred when opening,
+// reading, or closing either blob.
+func BlobEqual(ctx context.Context, a, b Blob) (_ bool, retErr error) {
+	if !a.Digest().Equal(*b.Digest()) {
+		// digests don't match
+		return false, nil
+	}
+	aFile, err := a.Open(ctx)
+	if err != nil {
+		return false, err
+	}
+	defer func() { retErr = multierr.Append(retErr, aFile.Close()) }()
+	bFile, err := b.Open(ctx)
+	if err != nil {
+		return false, err
+	}
+	defer func() { retErr = multierr.Append(retErr, bFile.Close()) }()
+	// Read blockSize from a, then from b, and compare.
+	aBlock := make([]byte, blockSize)
+	bBlock := make([]byte, blockSize)
+	for {
+		aN, aErr := aFile.Read(aBlock)
+		bN, bErr := io.ReadAtLeast(bFile, bBlock[:aN], aN) // exactly aN bytes
+		// We're running unexpected error processing (not EOF) before comparing
+		// bytes because it doesn't matter if the returned bytes match if an
+		// error occurred before an expected EOF.
+		if bErr == io.ErrUnexpectedEOF {
+			// b is shorter; we can error early
+			return false, nil
+		}
+		if aErr != nil && aErr != io.EOF {
+			// unexpected read error
+			return false, aErr
+		}
+		if bErr != nil && bErr != io.EOF {
+			// unexpected read error
+			return false, bErr
+		}
+		if !bytes.Equal(aBlock[:aN], bBlock[:bN]) {
+			// Read content doesn't match. Don't forget to forward any errors
+			// returned.
+			return false, multierr.Append(nilEOF(aErr), nilEOF(bErr))
+		}
+		if aErr == io.EOF || bErr == io.EOF {
+			// EOF
+			break
+		}
+	}
+	_, aErr := io.ReadAtLeast(aFile, aBlock, 1)
+	_, bErr := io.ReadAtLeast(bFile, bBlock, 1)
+	if aErr != io.EOF || bErr != io.EOF {
+		// a or b is longer, or we late errored
+		return false, multierr.Append(nilEOF(aErr), nilEOF(bErr))
+	}
+	return true, nil
```

Comment:
Couldn't we save the last check after the `for` loop, by checking both `aErr` and `bErr` are `io.EOF`?

```suggestion
		if aErr == io.EOF || bErr == io.EOF {
			// one of them is EOF...
			if aErr != bErr {
				// ...the other isn't
				return false, multierr.Append(nilEOF(aErr), nilEOF(bErr))
			}
			// ...the other too
			return true, nil
		}
	}
```

Reply:
Unfortunately not because each blob has a choice of returning EOF with data or data with no error then zero data and EOF. That's what that final reads are forcing: try really hard at reading the next byte.

This is somewhat confusing since io.ReadAtLeast complicates things even more. I really should just be calling Read directly and making the equal return a positive evaluation.